risc0/zkp/src/prove/soundness.rs

```diff
@@ -0,0 +1,292 @@
+        let epsilon_plus = 1.0 - rho_plus - theta;
+        let c_rho = 1; // unspecified exponent parameter in DEEP-FRI, Conjecture 2.3
+        (params.lde_domain_size / epsilon_plus).powi(c_rho)
+    };
+    soundness_error::<H>(taps, coeffs_size, theta, e_proximity_gap, l_plus)
+}
+
+fn soundness_error<H: Hal>(
+    taps: &TapSet,
+    coeffs_size: usize,
+    theta: f32,
+    e_proximity_gap: f32,
+    l_plus: f32,
+) -> f32 {
+    let params = parameters::<H>(taps, coeffs_size);
+    let plonk_plookup_error = params.plonk_plookup_error();
+    let fri_error = params.e_fri(theta, e_proximity_gap);
+    let deep_ali_error = params.e_deep_ali(l_plus);
+    libm::log2f(plonk_plookup_error + fri_error + deep_ali_error)
+}
+
+/// (1 - θ)^QUERIES
+fn e_fri_queries(theta: f32) -> f32 {
+    (1.0 - theta).powi(crate::QUERIES as i32)
+}
+
+/// Compute the number of folding rounds
+fn num_folding_rounds(coeffs_size: usize, ext_size: usize) -> usize {
+    let mut num_folding_rounds = 0;
+    let mut coeffs_size = coeffs_size;
+    while coeffs_size / ext_size > FRI_MIN_DEGREE {
+        coeffs_size /= FRI_FOLD;
+        num_folding_rounds += 1;
+    }
+    num_folding_rounds
+}
+
+#[derive(Copy, Clone)]
+struct Params {
+    /// The number of duplicated data columns in the witness that appear
+    /// due to the permutation from trace_time to trace_mem
+    n_sigma_mem: usize,
+    /// The number of columns in the witness that appear due to the bytes-lookup
+    n_sigma_bytes: usize,
+    /// No. of trace polynomials
+    n_trace_polys: f32,
+    /// Max degree of the constraint system, i.e. no. of segment polynomials
+    d: f32,
+    /// Max. no. of entries used from a single column
+    biggest_combo: f32,
+    /// Field extension degree
+    ext_size: usize,
+    /// Size of extension field
+    ext_field_size: f32,
+    /// Domain size of the trace
+    trace_domain_size: f32,
+    /// Domain size after low-degree extension
+    lde_domain_size: f32,
+    /// Number of folding rounds in FRI
+    num_folding_rounds: usize,
+}
+
+fn parameters<H: Hal>(taps: &TapSet, coeffs_size: usize) -> Params {
+    // Circuit-specific info
+    // FIXME get from circuit instead of hard-coding
+    let n_sigma_mem = 5;
+    // FIXME get from circuit instead of hard-coding
+    let n_sigma_bytes = 15;
+    let n_trace_polys = {
+        use crate::adapter::{REGISTER_GROUP_ACCUM, REGISTER_GROUP_CODE, REGISTER_GROUP_DATA};
+
+        let w_accum = taps.group_size(REGISTER_GROUP_ACCUM) as f32;
+        let w_code = taps.group_size(REGISTER_GROUP_CODE) as f32;
+        let w_data = taps.group_size(REGISTER_GROUP_DATA) as f32;
+
+        w_accum + w_code + w_data
+    };
+    // Max degree of the constraint system, i.e. no. of segment polys
+    // FIXME get from circuit instead of hard-coding
+    let d = 4.0;
+
+    let biggest_combo = taps.combos().map(|combo| combo.size()).max().unwrap() as f32;
+
+    // e, field extension degree
```

Comment:
is this a reference to the variable name in one of the other versions of this analysis?

Reply:
if so, i can revise the python notebook to use `ext_size`